domain/application/service/unit.go

```diff
@@ -152,29 +157,83 @@ func (s *Service) SetUnitPassword(ctx context.Context, unitName coreunit.Name, p
 // [applicationerrors.ApplicationNotFoundError] if the application doesn't
 // exist. If the unit life is Dead, an error satisfying
 // [applicationerrors.UnitAlreadyExists] is returned.
-func (s *Service) RegisterCAASUnit(ctx context.Context, appName string, args application.RegisterCAASUnitArg) error {
-	if args.PasswordHash == "" {
-		return errors.Errorf("password hash %w", coreerrors.NotValid)
+func (s *ProviderService) RegisterCAASUnit(
+	ctx context.Context, params application.RegisterCAASUnitParams,
+) (coreunit.Name, string, error) {
+	if params.ProviderID == "" {
+		return "", "", errors.Errorf("provider id %w", coreerrors.NotValid)
 	}
-	if args.ProviderID == "" {
-		return errors.Errorf("provider id %w", coreerrors.NotValid)
+
+	pass, err := password.RandomPassword()
+	if err != nil {
+		return "", "", errors.Errorf("generating unit password: %w", err)
 	}
-	if !args.OrderedScale {
-		return errors.Errorf("registering CAAS units not supported without ordered unit IDs").Add(coreerrors.NotImplemented)
+	registerArgs := application.RegisterCAASUnitArg{
+		ProviderID:       params.ProviderID,
+		StorageParentDir: application.StorageParentDir,
+		PasswordHash:     s.passwordHash(pass),
 	}
-	if args.UnitName == "" {
-		return errors.Errorf("missing unit name %w", coreerrors.NotValid)
+
+	// We don't support anything other that statefulsets.
+	// It's hard coded here until other deployment types are supported.
+	deploymentType := caas.DeploymentStateful
+	appName := params.ApplicationName
+
+	var unitName coreunit.Name
+	switch deploymentType {
+	case caas.DeploymentStateful:
+		splitPodName := strings.Split(params.ProviderID, "-")
+		ord, err := strconv.Atoi(splitPodName[len(splitPodName)-1])
+		if err != nil {
+			return "", "", errors.Capture(err)
+		}
+		if unitName, err = coreunit.NewNameFromParts(appName, ord); err != nil {
+			return "", "", errors.Capture(err)
+		}
+		registerArgs.UnitName = unitName
+		registerArgs.OrderedId = ord
+		registerArgs.OrderedScale = true
+	default:
+		return "", "", errors.Errorf("deployment type %q %w", deploymentType, coreerrors.NotSupported)
 	}
 
-	appUUID, err := s.st.GetApplicationIDByName(ctx, appName)
+	// Find the pod/unit in the provider.
+	k8sBroker, err := s.k8sBroker(ctx)
 	if err != nil {
-		return errors.Errorf("getting application ID: %w", err)
+		return "", "", errors.Errorf("registering k8s units for application %q: %w", appName, err)
 	}
-	err = s.st.RegisterCAASUnit(ctx, appUUID, args)
+	caasApp := k8sBroker.Application(appName, caas.DeploymentStateful)
+	pods, err := caasApp.Units()
```

Comment:
The modeling of this is wrong for me, but this is a long-standing issue with the caas provider. We shouldn't be leaking specific provider implementations (see caas package import) into the domain. We don't need to solve it now, and we can bounce it down the road, but ultimately, we don't ever leak lxd or aws into the application domain, nor should we be doing that for caas here.